# Biostats280 Homework 1
- Author: Shuang Gao 
- Date: 2018/04/17

## Question 2
2.1 Associative rule for addition says (x + y) + z == x + (y + z). Check association rule using x = 0.1, y = 0.1 and z = 1.0 in Julia. Explain what you find.

In [1]:
x, y, z = 0.1, 0.1, 1.0
(x + y) + z == x + (y + z)

false

In [2]:
@show (x + y) + z
@show x + (y + z)

(x + y) + z = 1.2
x + (y + z) = 1.2000000000000002


1.2000000000000002

- The floating point in the left hand side is rounded to the first decimal place, but the floating point in the right hand side is arounded to 16th decimal place. Such rounding difference results in the violation of the associative rule for floating point addition.

2.2 Do floating-point numbers obey the associative rule for multiplication: (x * y) * z == x * (y * z)?

In [3]:
x, y, z = 0.3, 0.9, 0.4
(x * y) * z == x * (y * z)

true

In [4]:
@show (x * y) * z
@show x * (y * z)

(x * y) * z = 0.10800000000000001
x * (y * z) = 0.10800000000000001


0.10800000000000001

In [5]:
x, y, z = 0.3, 0.1, 0.4
(x * y) * z == x * (y * z)

false

In [6]:
@show (x * y) * z
@show x * (y * z)

(x * y) * z = 0.012
x * (y * z) = 0.012000000000000002


0.012000000000000002

- Same thing happens as in part 2.1. The floating point in the left hand side is rounded to the third decimal place, but the floating point in the right hand side is arounded to 18th decimal place. Such rounding difference results in the violation of the associative rule for floating point multiplication.

2.3 Do floating-point numbers obey the distributive rule: a * (x + y) == a * x + a * y?

In [7]:
a, x, y = 0.4, 0.1, 0.3 
a * (x + y) == a * x + a * y 

false

In [8]:
@show a * (x + y)
@show a * x + a * y 

a * (x + y) = 0.16000000000000003
a * x + a * y = 0.16


0.16

- The floating point in the left hand side is rounded to the 17th  decimal place, but the floating point in the right hand side is arounded to second decimal place. Such rounding difference results in the violation of the distributive rule for floating point.

2.4 Is 0 * x == 0 true for all floating-point number x?

In [9]:
x = Inf
0 * x == 0

false

In [10]:
@show 0 * x

0x = NaN


NaN

- The result of 0 * Inf returns NaN, which means not a number. Thus the comparision between NaN and 0 is not valid, which returns a false boolean result.

2.5 Is x / a == x * (1 / a) always true? Try with x = 3.0, a = 5.0 and explain what you find.

In [11]:
x, a = 3.0, 5.0
x / a == x * (1 / a)

false

In [12]:
@show x / a 
@show x * (1 / a)

x / a = 0.6
x * (1 / a) = 0.6000000000000001


0.6000000000000001

- The left hand side result is rounded to the first decimal place, but the right hand side result is arounded to 16th decimal place. Thus, the rounding difference of division may result in the inequality of x / a == x * (1 / a).

## Quesiton 3


In [13]:
function g(k)
    for i = 1:10
        k = 5k - 1
    end
    k
end


g (generic function with 1 method)

3.1 Use @code_llvm to find the LLVM bitcode of compiled g with Int64 input.

In [14]:
a = 2
typeof(a)

Int64

In [15]:
@code_llvm g(a)
g(a)


; Function Attrs: uwtable
define i64 @julia_g_63890(i64) #0 !dbg !5 {
top:
  %1 = mul i64 %0, 9765625
  %2 = add i64 %1, -2441406
  ret i64 %2
}


17089844

3.2 Use @code_llvm to find the LLVM bitcode of compiled g with Float64 input.

In [16]:
b = 2.0
typeof(b)

Float64

In [17]:
@code_llvm g(b)


; Function Attrs: uwtable
define double @julia_g_63911(double) #0 !dbg !5 {
top:
  %1 = fmul double %0, 5.000000e+00
  %2 = fadd double %1, -1.000000e+00
  %3 = fmul double %2, 5.000000e+00
  %4 = fadd double %3, -1.000000e+00
  %5 = fmul double %4, 5.000000e+00
  %6 = fadd double %5, -1.000000e+00
  %7 = fmul double %6, 5.000000e+00
  %8 = fadd double %7, -1.000000e+00
  %9 = fmul double %8, 5.000000e+00
  %10 = fadd double %9, -1.000000e+00
  %11 = fmul double %10, 5.000000e+00
  %12 = fadd double %11, -1.000000e+00
  %13 = fmul double %12, 5.000000e+00
  %14 = fadd double %13, -1.000000e+00
  %15 = fmul double %14, 5.000000e+00
  %16 = fadd double %15, -1.000000e+00
  %17 = fmul double %16, 5.000000e+00
  %18 = fadd double %17, -1.000000e+00
  %19 = fmul double %18, 5.000000e+00
  %20 = fadd double %19, -1.000000e+00
  ret double %20
}


3.3 Compare the bitcode from questions 1 and 2. What do you find?

Let's reduce the loop in g function from 10 iterations to 3, and name it f, in order to find the way julia calculate.

In [18]:
function f(k)
    for i = 1:3
        k = 5k - 1
    end
    k
end


f (generic function with 1 method)

### Int64 input

In [19]:
# previous int64 input a
@code_llvm f(a)
f(a)


; Function Attrs: uwtable
define i64 @julia_f_63923(i64) #0 !dbg !5 {
top:
  %1 = mul i64 %0, 125
  %2 = add i64 %1, -31
  ret i64 %2
}


219

* From the output, julia attempts to calculate the result by modifying the equation $5k-1$ with associative and distributive rules first, and then plug in the unknown variable $k$. 
* Let's verify this idea with 3 iteration case:
$$
5(5(5k - 1) - 1) - 1 = 125k + (-25 -5 -1) \\
= 125k - 31
$$
125 and -31 are the results shown in `@code_llvm f(a)`. This modification can be done because integers obey associative and distributive rules in julia. 
* For function g, the calculation with Int64 input is modified to $9765625k-2441406$ with $k=a=2$, which shown in `@code_llvm g(a)`.

### Float64 input

In [20]:
# previous Float64 input b
@code_llvm f(b)
f(b)


; Function Attrs: uwtable
define double @julia_f_63925(double) #0 !dbg !5 {
top:
  %1 = fmul double %0, 5.000000e+00
  %2 = fadd double %1, -1.000000e+00
  %3 = fmul double %2, 5.000000e+00
  %4 = fadd double %3, -1.000000e+00
  %5 = fmul double %4, 5.000000e+00
  %6 = fadd double %5, -1.000000e+00
  ret double %6
}


219.0

* From the output, julia attempts to do the calculation following the iteration. For each iteration, the unknown $k$ is plugged in the equation $5k-1$. 
* Let's verify the idea with 3 iteration case:
 - For the first iteration, floating point multiplication is used for evaluatiing $5k-1$ with $k = 2.0$ and new $k=5k-1$ is stored. Thus, the parameters 5 and -1 are shown as floating points in `@code_llvm f(b)`.
 - For the second iteration, floating point multiplication is used for evaluatiing $5k-1$ with $k = 5*2.0-1$ and new $k=5k-1$ is stored. Thus, the parameters 5 and -1 are shown as floating points in `@code_llvm f(b)`.
 - For the third iteration, floating point multiplication is used for evaluatiing $5k-1$ with $k = 5*(5*2.0-1)-1$ and new $k=5k-1$ is stored. Thus, the parameters 5 and -1 are shown as floating points in `@code_llvm f(b)`.
* For function g, the results in `@code_llvm g(b)` show 10 pairs of parameters 5 and -1, which are shown as floating points. This reflects that for each iteration the variable $k$ is evaluated with floating point multiplication and addtion. Because floating point numbers are not necessary obeying the associative rule and distributive rule. The modified algorithm for integer numbers will not work for floating point case.


3.4 Read Julia documentation on @fastmath and repeat the questions 1-3 on the function

### Julia documentation

In [21]:
function g_fastmath(k)  
    @fastmath for i = 1:10  
        k = 5k - 1
    end
    k
end

g_fastmath (generic function with 1 method)

In [22]:
? @fastmath

No documentation found.

`Base.FastMath.@fastmath` is a macro.

```
# 1 method for macro "@fastmath":
@fastmath(expr::ANY) in Base.FastMath at fastmath.jl:127
```


* From the Julia documentation, @fastmath is a macro to allow floating point optimizations that are correct for real numbers, but lead to differences for IEEE numbers. This means, associative rule and distributive rule will work for floating numbers under @fastmath. 
* If this is the case, the float64 input for function g_fastmath will follow the same optimization as int64 input. This is, using distributive and associative rules to get the final expression with unknown $k$ and then plug in $k$.

### Int64

In [23]:
# previous int64 input a
@code_llvm g_fastmath(a)
g_fastmath(a)


; Function Attrs: uwtable
define i64 @julia_g_fastmath_64031(i64) #0 !dbg !5 {
top:
  %1 = mul i64 %0, 9765625
  %2 = add i64 %1, -2441406
  ret i64 %2
}


17089844

### Float64

In [24]:
# previous float64 input b
@code_llvm g_fastmath(b)
g_fastmath(b)


; Function Attrs: uwtable
define double @julia_g_fastmath_64033(double) #0 !dbg !5 {
top:
  %1 = fmul fast double %0, 9.765625e+06
  %2 = fadd fast double %1, -2.441406e+06
  ret double %2
}


1.7089844e7

* The LLVM bitcode for Int64 and for Float64 shows the same result, which means both of them optimize the code with distributive and associative rules first. 

## Question 4

Create the vector x = (0.988, 0.989, 0.990, ..., 1.010, 1.011, 1.012).

In [25]:
x = collect(0.988:0.001:1.012)

25-element Array{Float64,1}:
 0.988
 0.989
 0.99 
 0.991
 0.992
 0.993
 0.994
 0.995
 0.996
 0.997
 0.998
 0.999
 1.0  
 1.001
 1.002
 1.003
 1.004
 1.005
 1.006
 1.007
 1.008
 1.009
 1.01 
 1.011
 1.012

4.1 Plot the polynomial y = x^7 - 7x^6 + 21x^5 - 35x^4 + 35x^3 - 21x^2 + 7x -1 at points x.

In [ ]:
Pkg.add("Plots")
using Plots

In [ ]:
y = map(x -> x^7 - 7x^6 + 21x^5 - 35x^4 + 35x^3 
    - 21x^2 + 7x - 1, x)

In [ ]:
#Pkg.add("GR")
gr()
plot(x, y, label = "line")
scatter!(x, y, label = "points")
title!("Plot with expansion formula")

4.2 Plot the polynomial y = (x - 1)^7 at points x.

In [ ]:
y = map(x -> (x - 1)^7, x)

In [ ]:
gr()
plot(x, y, label = "line")
scatter!(x, y, label = "points")

4.3 Explain what you found.

Plot in part 2 is more smooth than that in part 1. The noise along the graph in part 1 is caused by the floating number calculation error. From Question 2, it is shown that for floating number calculation will induce some rounding errors. For equation in part 1, more floating number calculations are involved, thus more rounding errors show up and make the plot less smoothing.

### Question 5

Let the  n×n  matrix H have elements H[i, j] = 1 / (i + j - 1).

5.1 Write a function h(n) that outputs  n×n  matrix H. Try at least 3 ways, e.g., looping, comprehension, and vectorization. Compute and print H for n = 5.

### method 1: looping

In [ ]:
function h1(n)
    H = zeros(n, n)
    for i in 1:n, j in 1:n
        H[i, j] = 1 / (i + j - 1)
    end
    H
end
h1(5)

### method 2: comprehension

In [ ]:
function h2(n)
    H = [1 / (i + j - 1) for i in 1:n, j in 1:n]
end
h2(5)

### Method 3: vectorization

In [ ]:
function h3(n)
    H = 1 ./ (repmat(1:n, 1, n) + repmat((1:n)', n, 1) - 1)  
end
h3(5)  

5.2 Compare their efficiencies at n = 1000.

In [ ]:
Pkg.add("BenchmarkTools")
using BenchmarkTools
@benchmark h1(1000)

In [ ]:
@benchmark h2(1000)

In [ ]:
@benchmark h3(1000)

* Comparing the mean run time and GC for all three methods, looping and comprehension have relatively same run time and GC, around 6.5 ms and 17% GC, which are much more efficient than vectorization method at n = 1000. The allocation times are 2 times for looping and comprehension method, but around 6000 times for vectorization method, which further indicates looping and comprehension methods are much more efficient than vectorization method. 

5.3 Try different rounding modes (use function setrounding(Float64, RoundingMode)) and report the entry inv(H)[1, 1] for n = 15.

 From [julia documentation](https://docs.julialang.org/en/stable/stdlib/math/#Base.Rounding.RoundingMode), there are 7 rouding modes are curenttly supported. Here, let's try 4 different rounding modes:

#### 1. RoundNearest(default)

In [ ]:
H = h2(15)
setrounding(Float64, RoundNearest)
inv(H)[1, 1]

#### 2. RoundToZero

In [ ]:
setrounding(Float64, RoundToZero)
inv(H)[1, 1]

#### 3. RoundUp

In [ ]:
setrounding(Float64, RoundUp)
inv(H)[1, 1]

#### 4. RoundDown

In [ ]:
setrounding(Float64, RoundDown)
inv(H)[1, 1]

- From the results with different rounding modes, the entry (1, 1) ranges from round 149 to 157. Thus, the rounding error does affect the calculation results.

### Question 6

Let's show proof for part 3 first:

6.3 Show the **binomial inversion formula**
$$
	(\mathbf{A} + \mathbf{U} \mathbf{B} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{B}^{-1} + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1},
$$
where $\mathbf{A}$ and $\mathbf{B}$ are nonsingular.

Proof:
To show the **binomial inversion formula** is to show 
$$(\mathbf{A}+\mathbf{U} \mathbf{B} \mathbf{V}^T) \big\{ \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} ( \mathbf{B}^{-1}+ \mathbf{V}^T \mathbf{A}^{-1}\mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1}\big\}=\mathbf{I}
$$
$$
\begin{align}
& (\mathbf{A}+\mathbf{U} \mathbf{B} \mathbf{V}^T) \big\{ \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} ( \mathbf{B}^{-1}+ \mathbf{V}^T \mathbf{A}^{-1}\mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1}\big\} \\
= {}& \big\{ \mathbf{I}-\mathbf{U}(\mathbf{B}^{-1}+\mathbf{V}^T\mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1} \big\} +  \big\{ \mathbf{U} \mathbf{B}\mathbf{V}^T \mathbf{A}^{-1} - \mathbf{U} \mathbf{B} \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U} (\mathbf{B}^{-1}+ \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U} )^{-1} \mathbf{V}^T \mathbf{A}^{-1}\big\} \\
= {}& \big\{  \mathbf{I} + \mathbf{U} \mathbf{B} \mathbf{V}^T \mathbf{A}^{-1} \big\} - \big \{ \mathbf{U}(\mathbf{B}^{-1}+\mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1} + \mathbf{U}\mathbf{B}\mathbf{V}^T \mathbf{A}^{-1} \mathbf{U} (\mathbf{B}^{-1}+\mathbf{V}^T \mathbf{A}^{-1}\mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1} \big\} \\
= {}& \mathbf{I} + \mathbf{U}\mathbf{B}\mathbf{V}^T \mathbf{A}^{-1} - (\mathbf{U}+\mathbf{U}\mathbf{B}\mathbf{V}^T\mathbf{A}^{-1}\mathbf{U}) (\mathbf{B}^{-1} + \mathbf{V}^T \mathbf{A}^{-1}\mathbf{U})^{-1}\mathbf{V}^T \mathbf{A}^{-1} \\
= {}& \mathbf{I} + \mathbf{U}\mathbf{B}\mathbf{V}^T \mathbf{A}^{-1} - \mathbf{U}\mathbf{B} (\mathbf{B}^{-1}+ \mathbf{V}^T\mathbf{A}^{-1}\mathbf{U})(\mathbf{B}^{-1} + \mathbf{V}^T\mathbf{A}^{-1}\mathbf{U})^{-1}\mathbf{V}^T\mathbf{A}^{-1} \\
= {}& \mathbf{I} + \mathbf{U}\mathbf{B}\mathbf{V}^T\mathbf{A}^{-1} - \mathbf{U}\mathbf{B}\mathbf{V}^T\mathbf{A}^{-1} \\
= {}& \mathbf{I}
\end{align}
$$
Therefore, $
	(\mathbf{A} + \mathbf{U} \mathbf{B} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{B}^{-1} + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1}
$

6.1 Show the **Sherman-Morrison formula**
$$
	(\mathbf{A} + \mathbf{u} \mathbf{u}^T)^{-1} = \mathbf{A}^{-1} - \frac{1}{1 + \mathbf{u}^T \mathbf{A}^{-1} \mathbf{u}} \mathbf{A}^{-1} \mathbf{u} \mathbf{u}^T \mathbf{A}^{-1},
$$
where $\mathbf{A} \in \mathbb{R}^{n \times n}$ is nonsingular and $\mathbf{u} \in \mathbb{R}^n$. This formula supplies the inverse of the symmetric, rank-one  perturbation of $\mathbf{A}$.

Proof:

Let $\mathbf{B}$ be a 1 by 1 unit matrix, $\mathbf{V}=\mathbf{u}$ and $\mathbf{U}=\mathbf{u}$, 


$$
\begin{align}
& \because \mathbf{A} \in \mathbb{R}^{n \times n} \space\space\space \mathbf{u} \in \mathbb{R}^n \\
& \therefore \mathbf{u}^T\mathbf{A}^{-1}\mathbf{u} \text{ is a 1 by 1 unit matrix} \\
\end{align}
$$
By binomial inversion formula, 
$$
\begin{align}
& (\mathbf{A}+\mathbf{u} \mathbf{u}^T)^{-1} \\
= & \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{u} ( 1^{-1}+ \mathbf{u}^T \mathbf{A}^{-1}\mathbf{u})^{-1} \mathbf{u}^T \mathbf{A}^{-1} \\
= & \mathbf{A}^{-1} - \frac{1}{1 + \mathbf{u}^T \mathbf{A}^{-1} \mathbf{u}} \mathbf{A}^{-1} \mathbf{u} \mathbf{u}^T \mathbf{A}^{-1}\\
\end{align}
$$





6.2 Show the **Woodbury formula**
$$
	(\mathbf{A} + \mathbf{U} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1},
$$
where $\mathbf{A} \in \mathbb{R}^{n \times n}$ is nonsingular, $\mathbf{U}, \mathbf{V} \in \mathbb{R}^{n \times m}$, and $\mathbf{I}_m$ is the $m \times m$ identity matrix. In many applications $m$ is much smaller than $n$. Woodbury formula generalizes Sherman-Morrison and is valuable because the smaller matrix $\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U}$ is cheaper to invert than the larger matrix $\mathbf{A} + \mathbf{U} \mathbf{V}^T$.

Proof:

Let $\mathbf{B}=\mathbf{I}_m $, 

By binomial inversion formula,
$$
\begin{align}
& (\mathbf{A} + \mathbf{U} \mathbf{B} \mathbf{V}^T)^{-1} \\
= {}& (\mathbf{A} + \mathbf{U} \mathbf{I}_m \mathbf{V}^T)^{-1} \\
= {}& \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{I}_m^{-1} + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1}
\end{align}
$$

$$
\begin{align}
\because {}& (\mathbf{A} + \mathbf{U} \mathbf{V}^T)^{-1} = (\mathbf{A} + \mathbf{U} \mathbf{I}_m \mathbf{V}^T)^{-1} \\
\therefore {}& (\mathbf{A} + \mathbf{U} \mathbf{V}^T)^{-1} = \mathbf{A}^{-1} - \mathbf{A}^{-1} \mathbf{U} (\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})^{-1} \mathbf{V}^T \mathbf{A}^{-1}
\end{align}
$$

6.4 Show the identity
$$
	\text{det}(\mathbf{A} + \mathbf{U} \mathbf{V}^T) = \text{det}(\mathbf{A}) \text{det}(\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U}).
$$
This formula is useful for evaluating the density of a multivariate normal with covariance matrix $\mathbf{A} + \mathbf{U} \mathbf{V}^T$.

proof:

$$
\begin{align}
& \text{det}(\mathbf{A} + \mathbf{U} \mathbf{V}^T) \\
= {}& \text{det}{\mathbf{A}} + \text{det}{
\begin{pmatrix}
\mathbf{I_m} & \mathbf{V}^T \\
-\mathbf{A}^{-1}\mathbf{U} & \mathbf{I_n}
\end{pmatrix} 
} \\
= {}& \text{det}(\mathbf{A}) \text{det}(\mathbf{I}_m + \mathbf{V}^T \mathbf{A}^{-1} \mathbf{U})
\end{align}
$$